src/dump.c

```diff
@@ -2461,7 +2492,7 @@ static jl_method_t *jl_recache_method(jl_method_t *m)
 {
     assert(!m->is_for_opaque_closure);
     jl_datatype_t *sig = (jl_datatype_t*)m->sig;
-    jl_methtable_t *mt = jl_method_table_for((jl_value_t*)m->sig);
+    jl_methtable_t *mt = jl_method_get_table(m);
```

Comment:
@maleadt this is where it's supposed to see the other method table and re-insert it there (together with the code above that fetches the reference to the mt).

Reply:
The problem is that we never serialize (during precompilation) the overlay method when referring to a method table in another module, because we discover the methods to serialize by looking at what I assume are the internal method tables (`jl_type_type_mt` and `jl_nonfunction_mt`): https://github.com/JuliaLang/julia/blob/faa3d41f239adeef2692b2ad312a81f93fce8710/src/dump.c#L2204-L2219

Conversely, when the external method table is defined in the module that's being precompiled, we just serialize that object, and the fallback serialization that recurses into the method table's fields ends up serializing the overlay method.